src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,257 @@
+                                               gss_cred_id_t claimantCredHandle,
+                                               gss_ctx_id_t* contextHandle,
+                                               uint32_t isNtlm,
+                                               gss_name_t targetName,
+                                               uint32_t reqFlags,
+                                               uint8_t* inputBytes,
+                                               uint32_t inputLength,
+                                               gss_buffer_t* outBufferHandle,
+                                               uint32_t* outTokenLength,
+                                               uint32_t* retFlags)
+{
+    assert(isNtlm == 0 || isNtlm == 1);
+    assert(minorStatus != nullptr);
+    assert(contextHandle != nullptr);
+    assert(outBufferHandle != nullptr);
+    assert(outTokenLength != nullptr);
+    assert(retFlags != nullptr);
+#if HAVE_GSS_SPNEGO_MECHANISM
+    gss_OID desiredMech = isNtlm ? GSS_NTLM_MECHANISM : GSS_SPNEGO_MECHANISM;
+#else
+    assert(!isNtlm && "NTLM is not supported by MIT libgssapi_krb5");
+    char gss_mech_value[] = "\x2b\x06\x01\x05\x05\x02"; // Binary representation of SPNEGO Oid (RFC 4178)
+    gss_OID_desc gss_mech_spnego_OID_desc = {.length = 6, .value = static_cast<void*>(gss_mech_value)};
+    gss_OID desiredMech = &gss_mech_spnego_OID_desc;
+#endif
+
+    gss_buffer_desc inputToken{.length = UnsignedCast(inputLength), .value = inputBytes};
+    *outBufferHandle = new gss_buffer_desc();
+
+    uint32_t majorStatus = gss_init_sec_context(minorStatus,
+                                                claimantCredHandle,
+                                                contextHandle,
+                                                targetName,
+                                                desiredMech,
+                                                reqFlags,
+                                                0,
+                                                GSS_C_NO_CHANNEL_BINDINGS,
+                                                &inputToken,
+                                                nullptr,
+                                                *outBufferHandle,
+                                                retFlags,
+                                                nullptr);
+
+    NetSecurity_HandleError(majorStatus, outBufferHandle, outTokenLength);
+    return majorStatus;
+}
+
+extern "C" uint32_t NetSecurity_ReleaseCred(uint32_t* minorStatus, gss_cred_id_t* credHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(credHandle != nullptr);
+    return gss_release_cred(minorStatus, credHandle);
+}
+
+extern "C" uint32_t NetSecurity_ReleaseBuffer(uint32_t* minorStatus, gss_buffer_t buffer)
+{
+    assert(minorStatus != nullptr);
+    assert(buffer != nullptr);
+    uint32_t status = gss_release_buffer(minorStatus, buffer);
```

Comment:
> Doesn't the gss_release_buffer call free the memory in a gss_buffer_t?

It does not. It only frees the memory of the data it created. The buffer object itself (a structure with two fields) is owned by us.